internal/kv/tikv/txn_tikv.go

```diff
@@ -0,0 +1,805 @@
+func (kv *txnTiKV) MultiSaveAndRemoveWithPrefix(saves map[string]string, removals []string) error {
+	start := time.Now()
+	ctx, cancel := context.WithTimeout(context.Background(), RequestTimeout)
+	defer cancel()
+
+	var logging_error error
+	defer logWarnOnFailure(&logging_error, "txnTiKV MultiSaveAndRemoveWithPrefix() error", zap.Any("saves", saves), zap.Strings("removes", removals), zap.Int("saveLength", len(saves)), zap.Int("removeLength", len(removals)))
+
+	txn, err := beginTxn(kv.txn)
+	if err != nil {
+		logging_error = errors.Wrap(err, "Failed to create txn for MultiSaveAndRemoveWithPrefix")
+		return logging_error
+	}
+
+	// Defer a rollback only if the transaction hasn't been committed
+	defer rollbackOnFailure(&logging_error, txn)
+
+	// Save key-value pairs
+	for key, value := range saves {
+		key = path.Join(kv.rootPath, key)
+		// Check if value is empty or taking reserved EmptyValue
+		byte_value, err := convertEmptyStringToByte(value)
+		if err != nil {
+			logging_error = errors.Wrap(err, fmt.Sprintf("Failed to cast to byte (%s:%s) for MultiSaveAndRemoveWithPrefix()", key, value))
+			return logging_error
+		}
+		err = txn.Set([]byte(key), byte_value)
+		if err != nil {
+			logging_error = errors.Wrap(err, fmt.Sprintf("Failed to set (%s:%s) for MultiSaveAndRemoveWithPrefix()", key, value))
+			return logging_error
+		}
+	}
+	// Remove keys with prefix
+	for _, prefix := range removals {
+		prefix = path.Join(kv.rootPath, prefix)
+		// Get the start and end keys for the prefix range
+		startKey := []byte(prefix)
+		endKey := tikv.PrefixNextKey([]byte(prefix))
+
+		// Use Scan to iterate over keys in the prefix range
+		iter, err := txn.Iter(startKey, endKey)
+		if err != nil {
+			logging_error = errors.Wrap(err, fmt.Sprintf("Failed to create iterater for %s during MultiSaveAndRemoveWithPrefix()", prefix))
+			return logging_error
+		}
+
+		// Iterate over keys and delete them
+		for iter.Valid() {
+			key := iter.Key()
+			err = txn.Delete(key)
+			if logging_error != nil {
+				logging_error = errors.Wrap(err, fmt.Sprintf("Failed to delete %s for MultiSaveAndRemoveWithPrefix", string(key)))
+				return logging_error
+			}
+
+			// Move the iterator to the next key
+			err = iter.Next()
+			if err != nil {
+				logging_error = errors.Wrap(err, fmt.Sprintf("Failed to move Iterator after key %s for MultiSaveAndRemoveWithPrefix", string(key)))
+				return logging_error
+			}
+		}
+	}
+	err = kv.executeTxn(txn, ctx)
+	if err != nil {
+		logging_error = errors.Wrap(err, "Failed to commit for MultiSaveAndRemoveWithPrefix")
+		return logging_error
+	}
+	CheckElapseAndWarn(start, "Slow txnTiKV MultiSaveAndRemoveWithPrefix() operation", zap.Any("saves", saves), zap.Strings("removals", removals))
+	return nil
+}
+
+// WalkWithPrefix visits each kv with input prefix and apply given fn to it.
+func (kv *txnTiKV) WalkWithPrefix(prefix string, paginationSize int, fn func([]byte, []byte) error) error {
```

Comment:
use paginationSize as iterator batch or remove it?

Reply:
Done